casper: Case 1 for main_hash_voted_frac. So far error at z3 query.

denis-bogdanas opened the pull request.

daejunpark commented:
@denis-bogdanas Please use `;` for the comments for INI files.

denis-bogdanas commented:
Can you merge this pull request? At the moment main_hash_voted_frac() and total_curdyn_deposits_scaled() are completed and updated to latest Casper/EVM.

daejunpark commented on casper/casper-spec.ini:
```diff
@@ -44,6 +44,128 @@ requires:
+ensures:
+    ensures 0 <=Int RESULT                  andBool RESULT <=Int maxSInt128
+attribute:
+
```
@denis-bogdanas how about the negative cases for `total_curdyn_deposits_scaled`?

yzhang90 commented on Makefile:
```diff
@@ -210,31 +214,10 @@ $(specs_dir)/examples/sum-to-n-spec.k: resources/sum-to-n.md $(TANGLER)
-	@echo >&2 "==  gen-spec: $@"
-	mkdir -p $(dir $@)
-	python3 resources/gen-spec.py $^ delete_validator delete_validator > $@
+	python3 resources/gen-spec.py $^ $* $* > $@
```
It will not work for `vote` spec, because `vote` requires `recommended_target_hash` and `proc_reward`. You can not use `$*`.

denis-bogdanas commented:
But `recommended_target_hash-spec.k` and `proc_reward-spec.k` are also created.

daejunpark commented:
Note that the vote spec file (`vote-spec.k`) should contain the other two function's specs in the *same* file.

denis-bogdanas commented:
Ok, I'll see how to fix this makefile issue.

denis-bogdanas commented:
Can you do a code review now? `total_curdyn_deposits_scaled()` is completely proved. I also fixed the `Makefile`. Should satisfy requirements for `vote-spec.k`.

denis-bogdanas commented:
I forgot reverting the code on last line. Please revert it.

denis-bogdanas commented:
Added all public constant functions and increment_dynasty().

daejunpark commented:
@denis-bogdanas I cannot reproduce the result. Which versions of KEVM and K are you using? Also, what changes made to the bytecode?

daejunpark commented on casper/verification.k:
```diff
@@ -66,13 +66,30 @@ module VERIFICATION
+      requires A *Int B >Int maxUInt256
+
+    rule 0 -Word ((pow256 -Int chop( A *Int B )) /Int A) ==K B => false
+      requires A >=Int 0 andBool B >=Int 0 andBool chop(A *Int B) >Int maxSInt256
```
could you explain a bit about this? it's not trivial to me.

denis-bogdanas commented:
I just added proofs to the 2 lemmas in comments. Also added some requires clauses that were always true in evm context, but have to be here for correctness.

denis-bogdanas commented:
@daejunpark  Verified on latest versions of K and kevm that you suggested.
